packages/admin/cms-admin/src/userPermissions/hooks/currentUser.tsx

```diff
@@ -0,0 +1,54 @@
+import { gql, useQuery } from "@apollo/client";
+import { Loading } from "@comet/admin";
+import React from "react";
+
+import { ContentScopeInterface } from "../../contentScope/Provider";
+import { GQLCurrentUserPermission } from "../../graphql.generated";
+import { GQLCurrentUserQuery } from "./currentUser.generated";
+
+export type CurrentUserContext = { currentUser: CurrentUserInterface; isAllowed: (user: CurrentUserInterface, permission: string) => boolean };
+export const CurrentUserContext = React.createContext<CurrentUserContext | undefined>(undefined);
+
+export interface CurrentUserInterface {
+    name?: string;
+    email?: string;
+    language?: string;
+    permissions: GQLCurrentUserPermission[];
+    contentScopes: ContentScopeInterface[];
+}
+
+export const CurrentUserProvider: React.FC = ({ children }) => {
+    const { data, error } = useQuery<GQLCurrentUserQuery>(gql`
+        query CurrentUser {
+            currentUser {
+                name
+                email
+                contentScopes
+                permissions {
+                    permission
+                }
+            }
+            userPermissionsAvailablePermissions
+        }
+    `);
+
+    if (error) throw error.message;
+
+    if (!data) return <Loading behavior="fillPageHeight" />;
+
+    const context: CurrentUserContext = {
+        currentUser: data.currentUser,
+        isAllowed: (user: CurrentUserInterface, permission: string) => {
+            if (user.email === undefined) return false;
+            return user.permissions.some((p) => p.permission === permission);
+        },
+    };
+
+    return <CurrentUserContext.Provider value={context}>{children}</CurrentUserContext.Provider>;
+};
+
+export function useCurrentUser(): CurrentUserInterface {
+    const ret = React.useContext<CurrentUserContext | undefined>(CurrentUserContext);
+    if (!ret || !ret.currentUser) throw new Error("CurrentUser not found. Make sure CurrentUserContext exists.");
```

Comment:
How would this work while loading the user? Is this prevent by this https://github.com/vivid-planet/comet/pull/1360/files#diff-ae0d91d401713e1fe426795e8755cbddf9c8292a935b8dc08675191ca6203a67R37?

Reply:
Yes. That way useCurrentUser can always return a valid object.